src/io/mod.rs

```diff
@@ -1,5 +1,113 @@
 //! Input and output of images.
+
+use crate::{error, ImageError, ImageResult};
+
 mod reader;
 pub(crate) mod free_functions;
 
 pub use self::reader::Reader;
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+/// Set of supported strict limits for a decoder
+pub struct LimitSupport {
+    /// This field indicates whether the strict limit on the number of bytes is
+    /// supported.
+    pub max_alloc_strict: bool,
```

Comment:
Notwithstanding the comment on strictness of memory allocation above, I think there should be some way to distinguish between a best-effort enforcement and disregarding the limit entirely. Should this field be a tri-state bool or, in case strictness is deferred to another PR, be split into two?

Reply:
This may not have been entirely clear, but the idea was that you first set the limit and it is supported everywhere. However if you need to make sure that memory usage is limited for some reason you should be able to enable strict versions of limits. This means that if guaranteeing that usage is below the limit is not possible, decoding should fail, but *only* if the strict version is enabled. The best-effort enforcement is by default.

Reply:
What I was trying to say is, there are three versions of memory limit support:
* Decoders that don't care (the non-strict version here?). Using them to decode untrusted input is _never_ okay for a caller that is concerned about reliability of their programs.
* Decoders that haven't been written with a guarantee in mind, or that have been written to a slightly diverging specification of what the limit means, but will still use the limit as a guideline. A caller might decide that using those is fine with a decently large safety margin.
* Decoders that really _guarantee_ what we've specified that `strict` should mean.

Reply:
Couldn't the decoder set their own safety margin in that case? If they set an internal limit of 1GiB but might end up allocating 5GiB they might as well not have any limit. However on the other hand if we are reasonably sure that the are not going use more than 2x the limit it could first divide its own limit by 2 before setting it.

Thinking about it some more we may need to do some redesign of most (all?) of the limits of the various decoder to support the strict limits in way I was thinking. If so I am supportive of only having non-strict limits for now. I did add a simple check for `total_bytes` in `load` which should give us a little protection to begin with and adding other limits for the decoders we already have in a 

A way forward would probably be to introduce some sort of `limited-allocator` crate that we can integrate into the various decoders. And also have a shared `Limits` struct in this crate so all decoders can use a similar API. Am I correct in assuming that we could switch `image::io::Limits` to one reexported from another crate in a backwards-compatible way as long as we design the new one in an identical way?

Reply:
Not really, no. It's less about some factor but rather: The decoder calls into a decompression library or helper for X and it suspects that this lib will setup _some_ reasonable buffer for an operation but it won't or can't control the size of the buffer exactly. Then it can not argue for strict support but one can argue that it will reasonably ensure that the limit is adhered to. Note that it's not about a multiplicative margin but rather some fixed overhead that it might not have a lot of control over. (See also the stack-depth question).

> Am I correct in assuming that we could switch image::io::Limits to one reexported from another crate in a backwards-compatible way as long as we design the new one in an identical way?

No, strictly speaking we can't change a struct definition to a re-export without a breaking change as one might have local trait impls for both which would then conflict. We can however provide trivial converters as we do for `color_quant`.

Reply:
Note that the argument of unstrict or unknown library behavior already applies to `Vec::extend_from_slice` or `Vec::push`. Both do _some_ allocation and it very likely doubles the memory usage but we can't rely on it with a guarantee. Now that gets even more complicated for `HashMap`s or anything more complicated. It's also very hard to control exactly. A library would be indeed quite interesting to have for that use case.

Reply:
I quite like the idea of having each limit be a a tri-state bool. The non-strict variant could be called "best-effort" with a specification something like "exceeds the limit by no more than a small fixed (but unspecified) factor plus some additional constant amount".